syntax = "proto3";

package raftpb; 

service RaftService {
  // TODO: implement AppendEntries and RequestVote
  rpc RequestVote(RequestVoteRequest) returns (RequestVoteResponse);
  rpc AppendEntries(AppendEntriesRequest) returns (AppendEntriesResponse);
}

message RequestVoteRequest {
  int64 term = 1;
  int64 candidateId = 2;
  int64 lastLogIndex = 3;
  int64 lastLogTerm = 4;
}

message RequestVoteResponse {
  int64 term = 1;
  bool voteGranted = 2;
}

message AppendEntriesRequest {
  int64 term = 1;
  int64 leaderId = 2;
  int64 prevLogIndex = 3;
  int64 prevLogTerm = 4;
  repeated Entry entries = 5;
  int64 leaderCommit = 6;
}

message AppendEntriesResponse {
  int64 term = 1;
  bool success = 2;
}

message Entry {
  int64 term = 1;
  string command = 2;
}
